File: its-my-training-day/leetcode-top-75/questions/linked-list/linked-list/reorder-list/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const reorderList = (head: ListNode<number> | null): void => {
  if (head === null || head.next === null) return;
  let slow: ListNode<number> | null = head;
  let fast: ListNode<number> | null = head;
  while (fast?.next && fast.next.next) {
    slow = slow?.next ?? null;
    fast = fast.next.next;
  }
  const second = reverseLinkedList(slow?.next ?? null);
  if (slow) slow.next = null;
  let first = head;
  let secondNode = second;
  while (secondNode !== null) {
    const temp1 = first.next;
    const temp2 = secondNode.next;
    first.next = secondNode;
    secondNode.next = temp1;
    first = temp1 as ListNode<number>;
    secondNode = temp2;
  }
};

File: its-my-training-day/leetcode-top-75/questions/linked-list/linked-list/reorder-list/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def reorder_list(head: Optional[ListNode]) -> None:
    if not head or not head.next:
        return
    slow = head
    fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next
    second = reverse_linked_list(slow.next)
    slow.next = None
    first = head
    while second:
        temp1 = first.next
        temp2 = second.next
        first.next = second
        second.next = temp1
        first = temp1
        second = temp2
